modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryAdapter.java

```diff
@@ -654,12 +656,30 @@ private boolean isSafeLossPolicy() {
                 if (prj != null || part != null)
                     return nodes(cctx, prj, part);
 
-                if (cctx.affinityNode())
-                    return Collections.singletonList(cctx.localNode());
+                GridDhtPartitionTopology topology = cctx.topology();
 
-                Collection<ClusterNode> affNodes = nodes(cctx, null, null);
+                topology.readLock();
 
-                return affNodes.isEmpty() ? affNodes : Collections.singletonList(F.rand(affNodes));
+                try {
+                    if (cctx.affinityNode() && !topology.localPartitionMap().hasMovingPartitions())
```

Comment:
I suppose we can evaluate this branch without holding a topology read lock, cannot we?